### PR TITLE
Security score bar with email verification

### PR DESCRIPTION
**Pull request by lacksfish**
Added email verification to wallets UI
Added wallet security score bar which adjusts depending on the security score calculated.

Settings has been split into:
* Profile
* Security
* Tools

Please review, discuss.

### CODE REVIEW

File: src/js/app.js [hunk 1]
```diff
@@ -3,6 +3,7 @@ var blocktrail = angular.module("blocktrail.wallet", [
     "ui.bootstrap",
     "ui.bootstrap.dropdown",
     "ui.bootstrap.pagination",
+    "ui.bootstrap.tabs",
```
Comment (rubensayshi) [inline src/js/app.js]: afaik you're no longer using this right?

File: src/js/modules/core/controllers/verify-email/verify-email.ctrl.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        $scope.working  = true;
+        $scope.verified = false;
+        $scope.success  = false;
+
+        // Get state parameter
+        var token = $stateParams.token;
+
+        accountSecurityService.verifyEmail(token)
+            .then(function (result) {
+                if(result && result["data"] && !result["data"]["result"]) {
```
Comment (rubensayshi) [inline src/js/modules/core/controllers/verify-email/verify-email.ctrl.js]: NIT `result.data.result`

File: src/js/modules/core/controllers/verify-email/verify-email.ctrl.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+                $scope.working = false;
+            })
+            .catch(function () {
+                $scope.working = false;
+                return new Error($translate.instant("MSG_ERROR_TRY_AGAIN_LATER"))
+            });
+
+        $scope.continueToWallet = function() {
+            $location.path("/");
+        }
```
Comment (rubensayshi) [inline src/js/modules/core/controllers/verify-email/verify-email.ctrl.js]: don't use `$location` unless there's some desire need for this, use `$state.go`

File: src/js/modules/core/services/account-security-service/account-security.service.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            });
+        });
+
+        function updateSecurityScore() {
+            var settings = settingsService.getReadOnlySettingsData();
+            var score = 0.35 * settings.verifiedEmail + 0.35 * (settings.passwordScore / 4);
+
+            return launchService.getAccountInfo().then(function (accountInfo) {
+                if (accountInfo.requires2FA) {
+                    score += 0.3 * accountInfo.requires2FA
```
Comment (rubensayshi) [inline src/js/modules/core/services/account-security-service/account-security.service.js]: `0.3 * true = 0.3` <3 javascript

File: src/js/modules/setup/services/login/login.service.js [hunk 5]
```diff
@@ -110,9 +116,9 @@
         return self._launchService.storeAccountInfo(accountInfo)
             .then(function() {
                 return self._setupService.setUserInfo({
-                    username: data.responseData.username,
-                    displayName: data.responseData.username,
-                    email: data.responseData.email
+                    username:       data.responseData.username,
+                    displayName:    data.responseData.username,
+                    email:          data.responseData.email
```
Comment (rubensayshi) [inline src/js/modules/setup/services/login/login.service.js]: we're trying to get rid of this style, don't add more of it ;)

File: src/js/modules/wallet/configs/wallet-states/wallet-states.config.js [hunk 6]
```diff
@@ -78,13 +88,51 @@
                 }
             })
             /*--- Settings ---*/
+            // .state("app.wallet.settings", {
+            //     url: "/settings",
+            //     abstract: true,
+            //     controller: "SettingsCtrl"
+            // })
+
```
Comment (rubensayshi) [inline src/js/modules/wallet/configs/wallet-states/wallet-states.config.js]: remove dead code

File: src/js/modules/wallet/controllers/send-confirm-modal/send-confirm-modal.tpl.html [hunk 7]
```diff
@@ -74,7 +74,17 @@ <h3 class="modal-title">{{ 'SENDING_CONFIRM_TITLE' | translate }}</h3>
 
         <div ng-if="working && !complete">
             <div class="progress">
-                <div class="progress-bar" role="progressbar" aria-valuenow="0" aria-valuemin="0" aria-valuemax="100" style="width: {{progressWidth}}%;"></div>
+                <div class="progress-bar progress-bar-striped active"
+                     ng-class="{
+                         'progress-bar-warning': walletData.networkType == 'BTC',
+                         'progress-bar-success': walletData.networkType != 'BTC'
+                      }"
```
Comment (rubensayshi) [inline src/js/modules/wallet/controllers/send-confirm-modal/send-confirm-modal.tpl.html]: why only on BTC and not on BCH?

File: src/js/modules/wallet/controllers/settings/profile/settings.profile.tpl.html [hunk 8]
```diff
@@ -0,0 +1,178 @@
+<form class="form form-settings">
+    <!--<div class="alert" ng-class="{'alert-success': error.type=='success', 'alert-danger' : error.type=='danger'}">
+        {{ errMsg }}
+    </div>-->
```
Comment (rubensayshi) [inline src/js/modules/wallet/controllers/settings/profile/settings.profile.tpl.html]: dead code

File: src/js/modules/wallet/controllers/settings/security/settings.security.ctrl.js [hunk 9]
```diff
@@ -0,0 +1,513 @@
...
+    var settings = settingsService.getReadOnlySettingsData();
+
+    var listenerEnabled2faToggle;
+    // Final 2fa true/false
+    var isEnabled2fa = false;
+    // Toggle for enable in scope
+    $scope.isEnabled2fa = false;
+
+    // TODO: Show information about email verification
+    $scope.hasEmailChanged = false;
```
Comment (rubensayshi) [inline src/js/modules/wallet/controllers/settings/security/settings.security.ctrl.js]: unused

File: src/js/modules/wallet/controllers/settings/security/settings.security.ctrl.js [hunk 10]
```diff
@@ -0,0 +1,513 @@
...
+    formSettingsService.fetchData()
+        .then(initData);
+
+    /**
+     * Init data
+     *
+     * @param data
+     */
+    function initData(data) {
+
```
Comment (rubensayshi) [inline src/js/modules/wallet/controllers/settings/security/settings.security.ctrl.js]: we don't need a blank line at the start of every function (see other places as well)

File: src/js/modules/wallet/controllers/settings/security/settings.security.ctrl.js [hunk 11]
```diff
@@ -0,0 +1,513 @@
...
+                            });
+                    }, function() {
+                        // Reset for enter two factor token
+                        $scope.isEnabled2fa = isEnabled2fa;
+                    });
+            });
+    }
+
+    $scope.verifyEmail = function () {
+        var settings = settingsService.getReadOnlySettingsData();
```
Comment (rubensayshi) [inline src/js/modules/wallet/controllers/settings/security/settings.security.ctrl.js]: already got this at the top of the controller, can remove this line

File: src/js/modules/wallet/controllers/settings/security/settings.security.ctrl.js [hunk 12]
```diff
@@ -0,0 +1,513 @@
...
+        var sdk = activeWallet.getSdkWallet().sdk;
+
+        if (!settings.verifiedEmail) {
+            return sdk.requestVerificationEmail()
+                .then(function (result) {
+                    // err handling
+                    if (!result) {
+                        return dialogService.alert(
+                            "Oopss",
+                            "We've failed sending you a verification email to your email inbox."
```
Comment (rubensayshi) [inline src/js/modules/wallet/controllers/settings/security/settings.security.ctrl.js]: need some translations here and other places in this function

Comment (rubensayshi) [reply src/js/modules/wallet/controllers/settings/security/settings.security.ctrl.js]: also, errors should goto the error handler, not `!result`

File: src/js/modules/wallet/controllers/settings/settings.ctrl.js [hunk 13]
```diff
@@ -4,702 +4,5 @@
...
-                listenerFormSettings();
-            }
-
-            if(listenerEnabled2faToggle) {
-                listenerEnabled2faToggle()
-            }
-        });
-    }
-
+    function SettingsCtrl() { }
```
Comment (rubensayshi) [inline src/js/modules/wallet/controllers/settings/settings.ctrl.js]: @romaroid when controller is empty we can just omit it entirely?

**Comment by rubensayshi** [comment]
looks good, just some NITs

File: src/js/modules/core/services/account-security-service/account-security.service.js (hunk 4 above)
Comment (lacksfish) [reply src/js/modules/core/services/account-security-service/account-security.service.js]: aye ;)

File: src/js/modules/wallet/configs/wallet-states/wallet-states.config.js (hunk 6 above)
Comment (lacksfish) [reply src/js/modules/wallet/configs/wallet-states/wallet-states.config.js]: done

File: src/js/modules/setup/services/login/login.service.js (hunk 5 above)
Comment (lacksfish) [reply src/js/modules/setup/services/login/login.service.js]: done

File: src/js/modules/core/controllers/verify-email/verify-email.ctrl.js (hunk 3 above)
Comment (lacksfish) [reply src/js/modules/core/controllers/verify-email/verify-email.ctrl.js]: agree, should go to `app.wallet`

File: src/js/app.js (hunk 1 above)
Comment (lacksfish) [reply src/js/app.js]: true

File: src/js/modules/wallet/controllers/settings/security/settings.security.ctrl.js (hunk 9 above)
Comment (lacksfish) [reply src/js/modules/wallet/controllers/settings/security/settings.security.ctrl.js]: whoops yes, only belongs to profile ctrl

File: src/js/modules/wallet/controllers/settings/settings.ctrl.js (hunk 13 above)
Comment (lacksfish) [reply src/js/modules/wallet/controllers/settings/settings.ctrl.js]: removed the controller completely, as `app.wallet.settings` is now an abstract state.

File: src/js/modules/wallet/controllers/send-confirm-modal/send-confirm-modal.tpl.html (hunk 7 above)
Comment (lacksfish) [reply src/js/modules/wallet/controllers/send-confirm-modal/send-confirm-modal.tpl.html]: warning is yellow and it fits well with the blue for BTC. Yellow is also for the progress bar, but we can go green for every send-progress bar as well.

File: src/js/modules/core/controllers/verify-email/verify-email.ctrl.js [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+                    })
+                    .then(accountSecurityService.updateSecurityScore);
+
+            })
+            .then(function () {
+                $scope.working = false;
+            })
+            .catch(function () {
+                $scope.working = false;
+                return new Error($translate.instant("MSG_ERROR_TRY_AGAIN_LATER"))
```
Comment (rubensayshi) [inline src/js/modules/core/controllers/verify-email/verify-email.ctrl.js]: should throw, no return

File: src/js/modules/core/controllers/verify-email/verify-email.ctrl.js [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+                    $scope.success = true;
+                } else {
+                    $scope.success = false;
+                }
+
+                return settingsService.syncSettingsDown()
+                    .then(function(settingsReadOnlyData) {
+                        $scope.verified = settingsReadOnlyData.verifiedEmail;
+                    })
+                    .then(accountSecurityService.updateSecurityScore);
```
Comment (rubensayshi) [inline src/js/modules/core/controllers/verify-email/verify-email.ctrl.js]: don't think you should do this with "methods" of objects

File: src/js/modules/core/services/settings/settings.service.js [hunk 16]
```diff
@@ -573,13 +581,14 @@
...
                     glideraActivationNoticePending: self._doc.glideraActivationNoticePending,
+                    passwordScore: self._doc.passwordScore,
                     buyBTCRegion: self._doc.buyBTCRegion,
                     username: self._doc.username,
                     email: self._doc.email,
                     hideBCCSweepWarning: self._doc.hideBCCSweepWarning,
                     walletActivated: self._doc.walletActivated,
                     knownLanguages: self._doc.knownLanguages,
-                    showArchived: self._doc.showArchived
+                    showArchived: self._doc.showArchived,
```
Comment (rubensayshi) [inline src/js/modules/core/services/settings/settings.service.js]: shouldn't have trailing `,`

File: src/js/modules/wallet/controllers/settings/profile/settings.profile.ctrl.js [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+        profilePic: "/" + CONFIG.STATICSURL + "/img/blank_profile.png"
+    };
+
+    $scope.errors = {
+        email: false,
+        name: false
+    };
+
+    // $scope.languages = formSettingsService.getLanguages();
+    // $scope.currencies = formSettingsService.getCurrencies();
```
Comment (romaroid) [inline src/js/modules/wallet/controllers/settings/profile/settings.profile.ctrl.js]: uncomment languages and currencies and define them with empty array $scope.languages = [];

File: src/js/modules/wallet/controllers/settings/security/settings.security.ctrl.js [hunk 18]
```diff
@@ -0,0 +1,512 @@
...
+
+    $rootScope.pageTitle = 'SETTINGS';
+
+    // this automatically updates an already open modal instead of popping a new one open
+    // TODO remove it after moving modals change password, enable/disable 2FA
+    $scope.alert = dialogService.alertSingleton();
+
+    $scope.isLoading = true;
+
+    $scope.errors = {
```
Comment (romaroid) [inline src/js/modules/wallet/controllers/settings/security/settings.security.ctrl.js]: Remove $scope.errors we do not use it in this controller

File: src/js/modules/wallet/controllers/settings/security/settings.security.ctrl.js [hunk 19]
```diff
@@ -0,0 +1,512 @@
...
+                })
+                .catch(function (e) {
+                    return dialogService.alert(
+                        $translate.instant('ERROR_TITLE_2'),
+                        $translate.instant('MSG_ERROR_SENDING_MAIL')
+                    ).result;
+                });
+        } else {
+            return dialogService.alert(
+                "Nothing to do",
```
Comment (romaroid) [inline src/js/modules/wallet/controllers/settings/security/settings.security.ctrl.js]: Add to translations and replace with translation token

File: src/js/modules/core/services/account-security-service/account-security.service.js [hunk 20]
```diff
@@ -0,0 +1,69 @@
+(function () {
+    "use strict";
+
+    angular.module('blocktrail.core')
+        .factory('accountSecurityService', AccountSecurityService);
+
+    function AccountSecurityService(CONFIG, settingsService, $http, launchService) {
```
[the remark this answers is not in the document]
Comment (lacksfish) [reply src/js/modules/core/services/account-security-service/account-security.service.js]: agreed, thanks!

File: src/js/modules/wallet/controllers/settings/security/settings.security.ctrl.js (hunk 19 above)
Comment (lacksfish) [reply src/js/modules/wallet/controllers/settings/security/settings.security.ctrl.js]: thanks, missed this one

**Comment by lacksfish** [comment]
UI changes applied

**Pull request merged**